File: ros2/tutorials/CppROS2/src/cpp_pubsub/src/talker.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node {
public:
  MinimalPublisher()
  : Node("minimal_publisher"), count_(0) {
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
    timer_ = this->create_wall_timer(1s, [this]() { publish_message(); });
    RCLCPP_INFO(this->get_logger(), "Talker node started. Publishing to /chatter...");
  }

private:
  void publish_message() {
    std_msgs::msg::String msg;
    msg.data = "Hello World from C++: " + std::to_string(count_++);
    RCLCPP_INFO(this->get_logger(), "Publishing: [%s]", msg.data.c_str());
    publisher_->publish(msg);
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  size_t count_;
};

int main(int argc, char * argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MinimalPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: ros2/tutorials/RustROS2/install/rust_pubsub/share/rust_pubsub/rust/src/talker.rs
use anyhow::Result;
use rclrs::{Context, CreateBasicExecutor};
use std::time::Duration;
use std_msgs::msg::String as StringMsg;

fn main() -> Result<()> {
    // Initialize ROS and create a node through an executor (rclrs 0.7 API).
    let context = Context::default_from_env()?;
    let executor = context.create_basic_executor();
    let node = executor.create_node("minimal_publisher")?;

    let publisher = node.create_publisher::<StringMsg>("chatter")?;

    let mut count = 0;
    println!("Talker node started. Publishing to /chatter...");

    // 4. Main loop: publish every 1 second
    while context.ok() {
        let msg = StringMsg {
            data: format!("Hello World from Rust: {}", count),
        };
        
        println!("Publishing: [{}]", msg.data);
        publisher.publish(&msg)?;
        
        count += 1;
        std::thread::sleep(Duration::from_secs(1));
    }

    Ok(())
}